build-scripts/lerna-publish.js

```diff
@@ -0,0 +1,144 @@
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Custom lerna cli which adds a publication command to always attempt publication
+ * of all packages.
+ *
+ * Taken directly from the lerna publication command and lerna cli packages.
+ */
+
+'use strict';
+const cli = require('@lerna/cli');
+const addCmd = require('@lerna/add/command');
+const bootstrapCmd = require('@lerna/bootstrap/command');
+const changedCmd = require('@lerna/changed/command');
+const cleanCmd = require('@lerna/clean/command');
+const createCmd = require('@lerna/create/command');
+const diffCmd = require('@lerna/diff/command');
+const execCmd = require('@lerna/exec/command');
+const importCmd = require('@lerna/import/command');
+const initCmd = require('@lerna/init/command');
+const linkCmd = require('@lerna/link/command');
+const listCmd = require('@lerna/list/command');
+const publishCmd = require('@lerna/publish/command');
+const runCmd = require('@lerna/run/command');
+const versionCmd = require('@lerna/version/command');
+const Semver = require('semver');
+const pkg = require('lerna/package.json');
+const { PublishCommand } = require('@lerna/publish');
+const fs = require('fs');
+const path = require('path');
+
+function factory(argv) {
+  return new TravisPublishCommand(argv);
+}
+
+/** Override methods in the main publication command class to return the full set of packages for publication */
+class TravisPublishCommand extends PublishCommand {
+  findVersionedUpdates() {
+    let chain = Promise.resolve();
+
+    if (this.options.bump === "from-git") {
+      chain = chain.then(() => this.detectFromGit());
+    } else if (this.options.canary) {
+      chain = chain.then(() => this.detectCanaryVersions());
+    } else {
+      // If no version bump was specified, check to ensure the working directory is clean
+      // and then attempt to publish all packages.
+      chain = chain
+          .then(() => this.verifyWorkingTreeClean())
+          .then(() => Array.from(this.packageGraph.values()))
+          .then(updates => {
+            const updatesVersions = updates.map(({ pkg }) => [pkg.name, pkg.version]);
+
+            return {
+              updates,
+              updatesVersions,
+              needsConfirmation: true,
+            };
+          });
+    }
+
+    return chain;
+  }
+}
+
+// New command meta data
+const travisPublishCmd = Object.assign({}, publishCmd, {
+  command: 'publish-travis [bump]',
+  describe: 'Publish all packages in the current project',
+  handler: factory
+});
+
+// Setup the custom cli
+function main(argv) {
+  // For lerna publication to work, the NPM token must be stored in the .npmrc file in the user home directory
+  if (process.env.TRAVIS && process.env.NPM_TOKEN) {
+    fs.writeFileSync(
+        path.resolve(process.env.HOME, '.npmrc'),
+        `//registry.npmjs.org/:_authToken=${process.env.NPM_TOKEN}\n`,
+        'utf8');
+  }
+
+  const context = {
+    lernaVersion: pkg.version,
+  };
+
+  return cli()
```

Comment:
Please add comments.

Does each of these ultimately generate a shell command and execute it?
What are these commands doing, and why are they needed?

Are we building up a template script here that then gets filled in and executed by `.parse(argv, context)`?

Reply:
I can't answer a lot of these details other than to say this is how lerna setup it's CLI. I just copied it and added the custom travis command.

I did add a reference to the original source.

Reply:
Thanks, Chad.
I really don't want you to do any more research than you've already done.
I just want to capture whatever information / understanding you have for the benefit of future readers who have no experience with this stuff.

If you can just add 'this was copy/pasted from xyz' that is a big boost over no information at all.